### ISSUE TITLE
error when reload the module. 

### ISSUE THREAD

**Issue by agelin** [hansihe/Rustler#13]
Hello.Thanks for the great project.

I found the error when I reload nif module on iex.

```
The on_load function for module Elixir.Test returned {{:badmatch,
  {:error, {:upgrade, 'Upgrade not supported by this NIF library.'}}},
```

Is that has some way to add/edit/export the upgrade function?

Thank you.

**Comment by hansihe** [comment]
Yeah, I did not implement module reloading, as I don't currently see a way to maintain memory when it comes to resource structs across reloads.

I could add the option of allowing upgrades, but it would invalidate memory safety if you modify resource structs and reload the nif library. If you are not using resource structs, it would probably not be a problem.

**Comment by agelin** [comment]
Ok, thanks for your reply!

**Comment by hansihe** [comment]
I don't want to shut down the discussion around this or anything. If there is a need for this it should certainly be bumped up the feature priority list.

**Comment by agelin** [comment]
Last night I think about your anower. On reflection, I think you are right. 

Of cause,“upgrade” may be abused.

Close（/hide）some danger options, to keep the safe mode. It's always effective.

If possible, can there be more a clear full marco of insecurity？like “rustler_export_nifs!!!”，😄. developers can have more choices and know the consequences.

Thanks for the great job.

**Comment by jorendorff** [comment]
Upgrades probably can't be made safe. But we could probably add some kind of `unsafe upgrade: my_upgrade_fn` option to `rustler_export_nifs`. (Requiring the user to type in the `unsafe` keyword is good enough for me.)

**Comment by bigfish24** [comment]
Sorry if this is a basic question, but without support for NIF upgrades, what happens when you perform an upgrade?

**Comment by davydog187** [comment]
I believe the module will fail to load @bigfish24

**Comment by hansihe** [comment]
I believe what I wrote earlier in this thread is wrong.

As long as we always create resource types with the `ERL_NIF_RT_CREATE` flag, I believe we can do upgrades pretty safely.

Issues start to arise when we want to try supporting `ERL_NIF_RT_TAKEOVER`, since that's where we have the inability to prove that whatever is within the resource is identical between the two versions.

Having a basic version of NIF upgrading would certainly be better than having nothing at all
